apollo-router-core/src/spec/schema.rs

```diff
@@ -12,287 +14,318 @@ pub struct Schema {
-        for definition in document.definitions() {
-            macro_rules! implements_interfaces {
-                ($definition:expr) => {{
-                    let name = $definition
-                        .name()
-                        .expect("never optional according to spec; qed")
-                        .text()
-                        .to_string();
-
-                    for key in $definition
-                        .implements_interfaces()
-                        .iter()
-                        .flat_map(|member_types| member_types.named_types().flat_map(|x| x.name()))
-                    {
-                        let key = key.text().to_string();
-                        let set = subtype_map.entry(key).or_default();
-                        set.insert(name.clone());
-                    }
-                }};
+        fn parse(schema: &str) -> Result<Schema, SchemaError> {
+            let parser = apollo_parser::Parser::new(schema);
+            let tree = parser.parse();
+            let errors = tree.errors().cloned().collect::<Vec<_>>();
+
+            if !errors.is_empty() {
+                let errors = ParseErrors {
+                    raw_schema: schema.to_string(),
+                    errors,
+                };
+                errors.print();
+                return Err(SchemaError::Parse(errors));
             }
 
-            macro_rules! union_member_types {
-                ($definition:expr) => {{
-                    let key = $definition
-                        .name()
-                        .expect("never optional according to spec; qed")
-                        .text()
-                        .to_string();
-                    let set = subtype_map.entry(key).or_default();
-
-                    for name in $definition
-                        .union_member_types()
-                        .iter()
-                        .flat_map(|member_types| member_types.named_types().flat_map(|x| x.name()))
-                    {
-                        set.insert(name.text().to_string());
-                    }
-                }};
-            }
+            let document = tree.document();
+            let mut subtype_map: HashMap<String, HashSet<String>> = Default::default();
+            let mut subgraphs = HashMap::new();
+
+            // the logic of this algorithm is inspired from the npm package graphql:
+            // https://github.com/graphql/graphql-js/blob/ac8f0c6b484a0d5dca2dc13c387247f96772580a/src/type/schema.ts#L302-L327
+            // https://github.com/graphql/graphql-js/blob/ac8f0c6b484a0d5dca2dc13c387247f96772580a/src/type/schema.ts#L294-L300
+            // https://github.com/graphql/graphql-js/blob/ac8f0c6b484a0d5dca2dc13c387247f96772580a/src/type/schema.ts#L215-L263
+            for definition in document.definitions() {
+                macro_rules! implements_interfaces {
```

Comment:
Am I the only one feeling bad to see a macro_rules at this place ? I find this harder to read

Reply:
Agree, but it's out of the remit of this PR to change. It's just indenting that got modified here.